src/main/java/de/tum/in/www1/artemis/service/connectors/gitlabci/GitLabCIService.java

```diff
@@ -0,0 +1,399 @@
+        // Therefore, we return the key here even though GitLab CI does not need it.
+        return targetProjectKey + "-" + targetPlanName.toUpperCase().replaceAll("[^A-Z0-9]", "");
+    }
+
+    @Override
+    public void configureBuildPlan(ProgrammingExerciseParticipation participation, String defaultBranch) {
+        setupGitLabCIConfiguration(participation.getVcsRepositoryUrl(), participation.getProgrammingExercise(), participation.getBuildPlanId());
+    }
+
+    @Override
+    public void performEmptySetupCommit(ProgrammingExerciseParticipation participation) {
+        log.error("Unsupported action: GitLabCIService.performEmptySetupCommit()");
+    }
+
+    @Override
+    public void triggerBuild(ProgrammingExerciseParticipation participation) throws ContinuousIntegrationException {
+        triggerBuild(participation.getVcsRepositoryUrl(), participation.getProgrammingExercise().getBranch());
+    }
+
+    private void triggerBuild(VcsRepositoryUrl vcsRepositoryUrl, String branch) {
+        final String repositoryPath = getRepositoryPath(vcsRepositoryUrl);
+        try {
+            Trigger trigger = gitlab.getPipelineApi().createPipelineTrigger(repositoryPath, "Trigger build");
+            gitlab.getPipelineApi().triggerPipeline(repositoryPath, trigger, branch, null);
+            gitlab.getPipelineApi().deletePipelineTrigger(repositoryPath, trigger.getId());
+        }
+        catch (GitLabApiException e) {
+            throw new GitLabCIException("Error triggering the build for " + repositoryPath, e);
+        }
+    }
+
+    @Override
+    public void deleteProject(String projectKey) {
+        log.error("Unsupported action: GitLabCIService.deleteBuildPlan()");
+        log.error("Please refer to the repository for deleting the project. The build plan can not be deleted separately.");
+    }
+
+    @Override
+    public void deleteBuildPlan(String projectKey, String buildPlanId) {
+        log.error("Unsupported action: GitLabCIService.deleteBuildPlan()");
+        log.error("Please refer to the repository for deleting the project. The build plan can not be deleted separately.");
+    }
+
+    @Override
+    public String getPlanKey(Object requestBody) throws ContinuousIntegrationException {
+        TestResultsDTO dto = TestResultsDTO.convert(requestBody);
+        return dto.getFullName();
+    }
+
+    @Override
+    public AbstractBuildResultNotificationDTO convertBuildResult(Object requestBody) {
+        return TestResultsDTO.convert(requestBody);
+    }
+
+    @Override
+    public BuildStatus getBuildStatus(ProgrammingExerciseParticipation participation) {
+        // https://docs.gitlab.com/ee/api/pipelines.html#list-project-pipelines
+        final String repositoryPath = getRepositoryPath(participation.getVcsRepositoryUrl());
+        try {
+            // TODO: Get latest pipeline
+            PipelineStatus status = gitlab.getPipelineApi().getPipelines(repositoryPath).get(0).getStatus();
```

Comment:
This fetches commits from any branch. So if a student uses different branches to test stuff, then the wrong build status is shown.
Idea to solve the issue and the ToDo (untested):
```suggestion
        // https://docs.gitlab.com/ee/api/pipelines.html#list-project-pipelines
        try {
            final Optional<Pipeline> latestPipeline = getLatestPipeline(participation);
            if (latestPipeline.isEmpty()) {
                return BuildStatus.INACTIVE;
            }
```

with 

```java
    private Optional<Pipeline> getLatestPipeline(final ProgrammingExerciseParticipation participation) throws GitLabApiException {
        final String repositoryPath = getRepositoryPath(participation.getVcsRepositoryUrl());
        final Optional<String> commitHash = participation.findLatestSubmission().map(ProgrammingSubmission.class::cast).map(ProgrammingSubmission::getCommitHash);
        if (commitHash.isEmpty()) {
            return Optional.empty();
        }

        return gitlab.getPipelineApi().getPipelinesStream(repositoryPath, null, null, commitHash.get(), true, null, null,
            org.gitlab4j.api.Constants.PipelineOrderBy.ID, org.gitlab4j.api.Constants.SortOrder.DESC).findFirst();
    }
```